### PR TITLE
Implement RoaringBitmap.orNot operation #352

### PR DESCRIPTION
**Pull request by mdouaihy**
1. In-place orNot through a member method
2. non in-place orNot through a class method

### CODE REVIEW

**Comment by richardstartin** [comment]
Thanks for submitting these changes - it's very likely I will use this feature so want to help assure that it's correct and significantly outperforms alternatives already possible with the public API. I started looking at the changes this evening but encountered some issues running benchmarks the way I am used to. Unfortunately it will be a couple of days before I get more time to look at this.

**Comment by richardstartin** [comment]
@mdouaihy I took a look at the PR this afternoon. I refer to the evaluation code based on your branch [here](https://github.com/richardstartin/RoaringBitmap/tree/ornot-richard). I assume throughout that the implementation should satisfy the below, and compare your implementation to it.

```java
  static void verifyInvariance(RoaringBitmap left, RoaringBitmap right, long limit) {
    RoaringBitmap actual = RoaringBitmap.orNot(left, right, limit);
    RoaringBitmap range = new RoaringBitmap();
    range.add(0, limit);
    RoaringBitmap rightFlipped = RoaringBitmap.and(right, range);
    rightFlipped.flip(0, limit);
    RoaringBitmap expected = RoaringBitmap.or(RoaringBitmap.and(left, range), rightFlipped);
    assertEquals(expected, actual);
  }
```

1. I wrote a benchmark to compare your implementation with what a user could currently do, and found this PR gives a very large performance gain (maybe even 20x).
2. I wrote some basic scenario based tests which all passed.
3. I wrote a fuzz test and it didn't take long to find cases where the new function creates empty bitmaps. These seem to be limited to cases where the last value in the bitmap is signed negative. These cases are at `src/test/resources/testdata/fuzz/ornot` on my branch, but more can be generated by running `Fuzzer.orNot`.

So I think this is very good, but it looks like there is a signedness bug or two to track down prior to merging.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java [hunk 1]
```diff
@@ -1313,8 +1313,246 @@ public static RoaringBitmap andNot(final RoaringBitmap x1, final RoaringBitmap x
...
+      dest.highLowContainer.append(s2, Container.rangeOfOnes(0, lastSize + 1));
+    }
+  }
+
+  /**
+   * In-place bitwise ORNOT operation. The current bitmap is modified.
+   *
+   * @param other the other bitmap
+   */
+  public void orNot(final RoaringBitmap other, int rangeEnd) {
```
[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: indeed, will fix that.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java [hunk 2]
```diff
@@ -1313,8 +1313,246 @@ public static RoaringBitmap andNot(final RoaringBitmap x1, final RoaringBitmap x
...
+
+    int pos1 = 0, pos2 = 0;
+    int length1 = highLowContainer.size();
+    final int length2 = other.highLowContainer.size();
+
+    final int lastKey = Util.highbits(rangeEnd - 1);
```
[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: ok

[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: Indeed.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java [hunk 3]
```diff
@@ -1313,8 +1313,246 @@ public static RoaringBitmap andNot(final RoaringBitmap x1, final RoaringBitmap x
...
+    final int lastSize = Util.toIntUnsigned(Util.lowbits(rangeEnd - 1));
+
+    short s2 = 0;
+    while (pos1 < length1 && pos2 < length2 && s2 <= lastKey) {
```
[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: fixed. negative values taken into consideration with Util.compareUnsigned.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: fixed. the range is taken into consideration with Util.compareUnsigned.

[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: fixed.

File: RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java]: done.

**Comment by mdouaihy** [comment]
@richardstartin, I fixed the code according to your suggestions. 

I also integrated part of your code to enrich the Fuzzer tests. That allowed me to find some bugs, mainly at the edge of the universe (the last container at key = 65535).

I also integrated the benchmarks from your change in the PR.

File: RoaringBitmap/src/main/java/org/roaringbitmap/BitmapContainer.java [hunk 4]
```diff
@@ -8,6 +8,7 @@
 import org.roaringbitmap.buffer.MappeableContainer;
 
 import java.io.*;
+import java.lang.reflect.GenericDeclaration;
```
Comment (richardstartin) [inline RoaringBitmap/src/main/java/org/roaringbitmap/BitmapContainer.java]: optimise imports

File: fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java [hunk 5]
```diff
@@ -148,7 +148,14 @@ public static void verifyInvariance(String testName, int maxKeys, Predicate<Roar
               RoaringBitmap two = randomBitmap(maxKeys);
               if (validity.test(one, two)) {
                 try {
-                  Assert.assertEquals(left.apply(one, two), right.apply(one, two));
+                  T res1 = left.apply(one, two);
+                  T res2 = right.apply(one, two);
+                  if (!res1.equals(res2)) {
```
Comment (richardstartin) [inline fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java]: I'm not sure I understand the rationale for this change. Is this intentional? It's assumed that `left` is pure. If these isn't a good reason could you revert this please?

**Comment by richardstartin** [comment]
@mdouaihy I left a couple more comments but this looks very good to me. I looked at the build failure and it looks like travis being temperamental. Removing the accidental import will run the build again in any case.

@lemire this looks good to me - do you want to take a look?

File: fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java (hunk 5 above)
Comment (mdouaihy) [reply fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java]: Oops, I changed that code for debugging purposes. I forgot to reset the change. reverted.

File: RoaringBitmap/src/main/java/org/roaringbitmap/BitmapContainer.java (hunk 4 above)
Comment (mdouaihy) [reply RoaringBitmap/src/main/java/org/roaringbitmap/BitmapContainer.java]: done.

**Comment by lemire** [comment]
@richardstartin We are not porting this over to the `buffer` package? If it is not done within this PR, I think we should at least create a new issue for it.

From the look of it, this is a rather high level implementation, so a port to the `buffer` package would be easy.

Thoughts?

**Comment by mdouaihy** [comment]
@lemire, I pushed an implementation in the buffer package. As you said, the port was easy since it's the same algorithm.

File: fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java (hunk 5 above)
Comment (richardstartin) [reply fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java]: This isn’t resolved- the res1.equals(res2) line below also looks like debugging code. Can we revert this prior to merging please? 

FYI @lemire

**Comment by lemire** [comment]
Waiting for @richardstartin 's "second" approval before merging.

File: fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java (hunk 5 above)
Comment (mdouaihy) [reply fuzz-tests/src/test/java/org/roaringbitmap/Fuzzer.java]: @richardstartin, indeed, I lost the revert while pushing. should be ok now.

**Comment by richardstartin** [comment]
Looks really good 👍

**Comment by lemire** [comment]
Merging.

**Pull request merged**